tests/MongoDB.Driver.Tests/Jira/CSharp3188Tests.cs

```diff
@@ -0,0 +1,95 @@
+﻿/* Copyright 2020-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.IO;
+using System.Net.Sockets;
+using FluentAssertions;
+using MongoDB.Bson;
+using MongoDB.Bson.TestHelpers.XunitExtensions;
+using MongoDB.Driver.Core.Clusters;
+using MongoDB.Driver.Core.Misc;
+using MongoDB.Driver.Core.TestHelpers.XunitExtensions;
+using Xunit;
+
+namespace MongoDB.Driver.Tests.Jira
+{
+    public class CSharp3188Tests
+    {
+        [SkippableTheory]
+        [ParameterAttributeData]
+        public void Ensure_that_MongoConnectionException_contains_expected_attributes([Values(false, true)] bool async)
+        {
+            RequireServer
+                .Check()
+                .Supports(Feature.AggregateFunction)
+                .ClusterTypes(ClusterType.Standalone, ClusterType.ReplicaSet); // server sleep doesn't correctly work with sharded
+
+            var socketTimeout = TimeSpan.FromMilliseconds(100);
+            var mongoClientSettings = DriverTestConfiguration.GetClientSettings().Clone();
+            mongoClientSettings.SocketTimeout = socketTimeout;
+
+            using (var client = DriverTestConfiguration.CreateDisposableClient(mongoClientSettings))
+            {
+                var database = client.GetDatabase("db");
+                var collection = database.GetCollection<BsonDocument>("coll");
+
+                var serverResponseDelay = socketTimeout + TimeSpan.FromMilliseconds(500);
```

Comment:
this value is quite big, but it's the lowest value which correctly works on EG. The test results are quite unclear.
For example, if `serverResponseDelay =500`, https://evergreen.mongodb.com/task/dot_net_driver_secure_tests__version~4.4_os~windows_64_topology~standalone_auth~auth_ssl~ssl_test_net452_patch_e604deb755d600ef3a628c06f8cdfdf757d073d1_5f72470c0305b95a3533db27_20_09_28_20_26_52 the server logs is the following:

    async:false:
    `09.44.407` starting the aggregate command with waiting. The test failed since we expected the error. Message: `Expected type to be MongoDB.Driver.MongoConnectionException, but found <null>.`
    `09:44.834` the test is finished. No errors are thrown
    
    async:true
    `09:45.020` starting the aggregate command with waiting.
    `09:45.186` the test is finished. The expected error has been thrown  

The biggest question that enough delay was when `async=false`, so this test should be successful, but EG results are completely the opposite. The successful test was when `async=true` even though I don't see enough delay in the server log and as the result, this test should be without error and as the result should be failed. 

So, mostlikely, EG log contains the wrong test description. And the failed test was `async`, but sync was successful.
As for why we need so long a delay (600ms), I don't have a strong answer, I guess some server restriction related to how the server sleep works

Reply:
NOTE: if anyone will check server logs, the above patch used appName=`test_{async}`